Huggingface Integration

vwxyzjn opened the pull request:
## Description
This PR closes #110.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix
- [x] New feature
- [ ] New algorithm
- [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [ ] I have ensured `pre-commit run --all-files` passes (**required**).
- [ ] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [ ] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [ ] I have contacted [vwxyzjn](https://github.com/vwxyzjn) to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [ ] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [ ] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [ ] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [ ] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm.
    - [ ] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [ ] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [ ] I have added links to the tracked experiments.
    - [ ] I have updated the overview sections at the [docs](https://docs.cleanrl.dev/rl-algorithms/overview/) and the [repo](https://github.com/vwxyzjn/cleanrl#overview)
- [ ] I have updated the tests accordingly (if applicable).

vwxyzjn commented:
The integration also makes it easier to just run models, such as 

https://github.com/vwxyzjn/cleanrl/blob/4074eee3b785fe06b90e48b5abfd404e896d41b8/cleanrl_utils/evals/dqn_eval.py#L43-L57

vwxyzjn commented:
CC @ThomasSimonini for review :) Thanks!

kinalmehta commented on cleanrl_utils/evals/dqn_eval.py:
```diff
@@ -0,0 +1,59 @@
+
+    from cleanrl.dqn import QNetwork, make_env
+
+    model_path = hf_hub_download(repo_id="cleanrl/CartPole-v1-dqn-seed1", filename="q_network.pth")
```
It would be nice to have the ability to pass hf-entity and hf-repo name as an optional parameter if a user wants to evaluate model saved to their own repo.

vwxyzjn replied:
Thank you, Kinal. I was thinking of giving a reasonable preset as follows

```
            repo_name = f"{args.env_id}-{args.exp_name}-seed{args.seed}"
            repo_id = f"{args.hf_entity}/{repo_name}" if args.hf_entity else repo_name
```

and the user, of course, has the flexibility to customize the name of the repo through `exp_name` like

```
python dqn.py --exp-name my-custom-dqn
```

and the repo id will be f"{args.hf_entity}/CartPole-v1-my-custom-dqn-seed1"

kinalmehta reviewed (approved):
LGTM.

simoninithomas reviewed (commented):
First, I want to thank you for this integration and all the work behind 🤗. The result is fantastic, especially the model card with visible hyperparameters.

I gave some insights based on Omar and Lucain and my review on how we can improve the `push_to_hub` part. I'm happy to help with this.**

In addition to this, we can:
1. Having a downstream part ( `load_from_hub` ), I can help with that.
2. Generating a json/yaml file containing the hyperparameters for reproducibility?
3. Adding the library to our Hub list so that it creates a tag for people searching for cleanrl models.

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,94 @@
+    with open(readme_path, "w", encoding="utf-8") as f:
+        f.write(readme)
+    metadata_save(readme_path, metadata)
+    repo_url = upload_folder(
```
Hi @vwxyzjn, I'm Lucain working at Huggingface on the `huggingface_hub` library. I would encourage you as well to use the `CommitOperationAdd` and `CommitOperationDelete` operations so that you can push your model and have a single commit in the history on the Hub. It should be well enough documented but if you have any questions, do not hesitate to ask !

vwxyzjn commented:
@kinalmehta @simoninithomas and @Wauplin, thanks for the review. The `CommitOperation` suggestion is really helpful. Regarding some further comments:

> Having a downstream part ( load_from_hub ), I can help with that.
Appreciate the help! That said, we are only downloading a single file from the hub, so having a customized load_from_hub might be unnecessary, right?

https://github.com/vwxyzjn/cleanrl/blob/b4305405dbeabbbed6058b67b161b48325d094da/cleanrl_utils/evals/dqn_eval.py#L48

> Generating a json/yaml file containing the hyperparameters for reproducibility?
Are you thinking of loading from the yaml file somehow to run the script like `python dqn.py --load-yaml hyper.yaml`?


> Adding the library to our Hub list so that it creates a tag for people searching for cleanrl models.
That would be great! Thank you!

simoninithomas commented:
Hi @vwxyzjn , yes for yaml I was thinking what you mentioned.

> That said, we are only downloading a single file from the hub, so having a customized load_from_hub might be unnecessary, right?

Yes and no, because it has two advantages:
1. We are able to count how many download of the model each month.
![image](https://user-images.githubusercontent.com/13088472/196892427-484432c2-00fa-40b4-bd4e-89975df5cc2d.png)

2. We can cache the model without using `hf_hub_download` directly.

For instance with SB3 integration here's the code for load_from_hub:

```python
def load_from_hub(repo_id: str, filename: str) -> str:
    """
    Download a model from Hugging Face Hub.
    :param repo_id: id of the model repository from the Hugging Face Hub
    :param filename: name of the model zip file from the repository
    """
    try:
        from huggingface_hub import hf_hub_download
    except ImportError:
        raise ImportError(
            "You need to install huggingface_hub to use `load_from_hub`. "
            "See https://pypi.org/project/huggingface-hub/ for installation."
        )

    # Get the model from the Hub, download and cache the model on your local disk
    downloaded_model_file = hf_hub_download(
        repo_id=repo_id,
        filename=filename,
        library_name="huggingface-sb3",
        library_version="2.1",
    )

    return downloaded_model_file
```

simoninithomas commented:
FIY From our side, we started to work on the frontend integration 🤗 
https://github.com/huggingface/hub-docs/pull/447

vwxyzjn commented:
Thank you @simoninithomas 

> We are able to count how many download of the model each month.

Does this mean `hf_hub_download(repo_id="cleanrl/CartPole-v1-dqn-seed1", filename="q_network.pth")` would not trigger the download stats?

> We can cache the model without using hf_hub_download directly.

Does `hf_hub_download` not cache models? I ran the `dqn_eval.py` and noticed the download progress bar only presents once and it did not appear again during the second run, so I assumed `hf_hub_download` caches automatically.

> FIY From our side, we started to work on the frontend integration 🤗
https://github.com/huggingface/hub-docs/pull/447

Awesome thanks! :)

Wauplin commented:
Hi @vwxyzjn

> Does this mean hf_hub_download(repo_id="cleanrl/CartPole-v1-dqn-seed1", filename="q_network.pth") would not trigger the download stats?

I'll let @simoninithomas answer on that as I am 100% sure what is counted in # downloads / month.
However, [the example](https://github.com/vwxyzjn/cleanrl/pull/292#issuecomment-1285115114) from @simoninithomas uses 2 kwargs `library_name` and `library_version` to make the Hub know which lib is downloading the model (e.g. a `cleanrl` user and not a random user).

> Does hf_hub_download not cache models?

Yes it does ! No matter if you use `hf_hub_download` or `snapshot_download` , your files will be downloaded only once.

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,126 @@
+        if ".tfevents" in file or ".mp4" in file:
+            operations += [CommitOperationDelete(path_in_repo=file)]
+
+    api.create_commit(
+        repo_id=repo_id,
+        operations=operations,
+        commit_message="clean up previous tfevents and mp4 files",
+    )
```
If you want you can combine the deletions and the additions in a single commit. Before `huggingface_hub==0.11.0` release this was dangerous as order of the operations was not respected (first add, then delete). But now this issue has been fixed which means you can have a list of operations with `CommitOperationDelete` first and then the `CommitOperationAdd`.

Having a single commit can avoid a weird situation where files have been deleted but upload failed.

vwxyzjn replied:
Nice thanks for this suggestion.

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,126 @@
+
+    # Step 2: clean up data
+    # delete previous tfevents and mp4 files
+    operations = []
+    for file in api.list_repo_files(repo_id=repo_id):
+        if ".tfevents" in file or ".mp4" in file:
+            operations += [CommitOperationDelete(path_in_repo=file)]
```
(nit) I would use str.endswith("...") to check file extension

```suggestion
    operations = [
        CommitOperationDelete(path_in_repo=file)
        for file in api.list_repo_files(repo_id=repo_id)
        if file.endswith(".tfevents") or file.endswith(".mp4")
    ]
```

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,126 @@
+HUGGINGFACE_README_FILE_NAME = "README.md"
+
+
+def push_to_hub(
+    args: argparse.Namespace,
+    episodic_returns: List,
+    repo_id: str,
+    algo_name: str,
+    folder_path: str,
+    video_folder_path: str = "",
+):
```
(suggestion)

If you have a single `create_commit(...)` call, it can be easy to add two parameters:
- `revision: Optional[str] = None`: allows to push to a different branch on the repo. Push on `main` by default.
- `create_pr: bool = False`: allows to create a PR in case user is not pushing on its own repo.
Both params can be provided to `create_commit` without any other processing on `cleanrl` side.

Also possible to add a `private: bool = False` argument that can be passed to `create_repo(...)` to create a private repo instead.

vwxyzjn replied:
Done.

Wauplin reviewed (commented):
Hi @vwxyzjn,  I made a review only on the `huggingface` integration part. I've added a few suggestions to integrate even better with the Hub. To allow them, it would be necessary to pin the `huggingface_hub` version to `>=0.11` (see [recent release](https://github.com/huggingface/huggingface_hub/releases/tag/v0.11.0)).

Otherwise in general, looks good to me :)

vwxyzjn commented:
Hey, @simoninithomas I have also added an `enjoy.py` and added reproducibility information. Support for `dqn_jax.py` is also added. What do you think of the design? If all the design looks good, the next step is to run more experiments.

simoninithomas commented:
Sorry for the delay @vwxyzjn the model card design is super good 😍. 

What do you think are the next steps ? And how I can help?

I think from my side:
- Documentation update with update of the integration
- Colab tutorial finalize
- Blogpost announcing integration on HF to update https://github.com/huggingface/blog/pull/616

vwxyzjn commented:
@simoninithomas, thanks for your patience.

> Edit: The only thing is that from now, the Hub is not recognizing CleanRL lib (not generating a specific lib tag) so what I can do is add it to our UI first before your tests if you prefer.

Please give it a go :)

> In order to do that we need to add:
metadata["library_name"] = "cleanrl" during the model card generation (like this in SB3: https://github.com/huggingface/huggingface_sb3/blob/main/huggingface_sb3/push_to_hub.py#L184)

Done.

> I think from my side:
> * Documentation update with update of the integration
> * Colab tutorial finalize
> * Blogpost announcing integration on HF to update https://github.com/huggingface/blog/pull/616
> * Adding the lib to the Hub to generate the tags

All these sound good. I have just given you access to the repo in case you want to directly modify anything in this branch :) The current docs I have added are at https://github.com/vwxyzjn/cleanrl/blob/hf-integration/docs/get-started/zoo.md. 

On my side, I will work on creating a PyPi release with this PR to enable the *load models* stuff colab tutorial. However, training with colab is still probably tricky because the training code is in the `if __name__=="__main__"` block.


Three todo-items on my side:
- [ ] merge #320 and rebase the poetry files
- [ ] make a (test) PyPi release to enable model loading and evaluation on colab (this requires further playing with `poetry`)
- [ ] run `dqn`, `dqn_jax`, `dqn_atari` experiments on our standard benchmark environments.
- [ ] merge this PR and make an official PyPi release, along with https://github.com/huggingface/blog/pull/616

vwxyzjn commented:
https://colab.research.google.com/drive/1vhbb4ak9smE7-LiHdaxwTACi1oOC5a4l?usp=sharing has a preliminary demo. To make the extra options appear during the pypi installation I had to do some hacks as suggested in  https://github.com/python-poetry/poetry/issues/4842#issuecomment-1340462066

simoninithomas commented:
Awesome thanks @vwxyzjn for your work I'm going to test that this afternoon 🤗 keep you updated.

simoninithomas commented:
First of all, thank you very much for your work @vwxyzjn. I tried the collab. I have multiple questions:

1. You call the architectures: `exp_name` I was wondering if  it can be misleading 🤔 .
2. The idea is to add some variants from now and in the future have all the algorithms working with the load model / save model?
3. Thank you very much for the docs document it's good (I have some updates to do on this part but I'll ask you first). I need to verify with RL team and make some tests.

The model cards are very good. From my side 

To give you a timeframe, my idea is to be focused on this integration next week so don't hesitate to tell me how we can help you in addition to the tasks mentioned 🤗.

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,128 @@
+    from huggingface_hub import CommitOperationAdd, CommitOperationDelete, HfApi
+    from huggingface_hub.repocard import metadata_eval_result, metadata_save
+
+    api = HfApi()
```
```suggestion
    api = HfApi(library_name="cleanrl", library_version="1.0.0")
```

This is optional but helps us understand usage of `huggingface_hub` (e.g. "how many users from 'cleanrl' ?")

vwxyzjn replied:
This gives me 
```
Traceback (most recent call last):
  File "/home/costa/Documents/go/src/cleanrl/cleanrl/dqn_jax.py", line 263, in <module>
    push_to_hub(args, episodic_returns, repo_id, "DQN", f"runs/{run_name}", f"videos/{run_name}-eval")
  File "/home/costa/Documents/go/src/cleanrl/cleanrl_utils/huggingface.py", line 29, in push_to_hub
    api = HfApi(library_name="cleanrl", library_version="1.0.0")
TypeError: __init__() got an unexpected keyword argument 'library_name'
```

Wauplin replied:
Oops sorry. Available in next version as well so I'll include it in the follow-up PR :)

vwxyzjn replied:
Sounds good. Oh also a related problem — I noticed the replay.mp4 is not longer displayed at the model card https://huggingface.co/cleanrl/BreakoutNoFrameskip-v4-dqn_atari_jax-seed1. Would you mind checking if that's the case?

Wauplin replied:
Just for context, was it ever included in the `README.md` or do you mean the widget displayed on the right side of the model card (above inference API) ? Maybe @simoninithomas could have an insight on this.

vwxyzjn replied:
All good. It's just the other day the video in this screenshot is not showing, but it's working today.
<img width="1444" alt="Screenshot 2022-12-20 at 9 36 53 AM" src="https://user-images.githubusercontent.com/5555347/208692170-31a7fac9-4584-45d9-a554-8daad34ef5e2.png">

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,128 @@
+        exist_ok=True,
+        private=private,
+    )
+    # parse the default entity
+    entity, repo = repo_url.split("/")[-2:]
+    repo_id = f"{entity}/{repo}"
```
I'm really sorry this is not easier at the moment :confused: 
I made a PR to fix this but it will only be available after next release of `huggingface_hub`. I can open another PR on "cleanrl" once it's release if you want (and you can keep this manual parsing for now, that's fine).

vwxyzjn replied:
Thanks for the comment! Another PR once it's released would be good :)

Wauplin replied:
Will do, it's not urgent anyway :)

Wauplin commented on cleanrl_utils/huggingface.py:
````diff
@@ -0,0 +1,128 @@
+{pformat(vars(args))}
+```
+    """
+    readme_path = f"{folder_path}/README.md"
````
```suggestion
    readme_path = Path(folder_path) / HUGGINGFACE_README_FILE_NAME
```
Here it would break for your Windows users. Also you can use the `HUGGINGFACE_README_FILE_NAME` variable you defined above :)

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,128 @@
+    metadata_save(readme_path, metadata)
+
+    # fetch mp4 files
+    if video_folder_path:
+        video_files = [str(item) for item in Path(video_folder_path).glob("*.mp4")]
+        # sort by the number in the file name
+        video_files = sorted(video_files, key=lambda x: int("".join(filter(str.isdigit, os.path.splitext(x)[0]))))
+        for file in video_files:
+            operations += [CommitOperationAdd(path_or_fileobj=file, path_in_repo=file)]
+        operations += [CommitOperationAdd(path_or_fileobj=file, path_in_repo=HUGGINGFACE_VIDEO_PREVIEW_FILE_NAME)]
```
```suggestion
    if video_folder_path:
        # Push all video files
        video_files = list(Path(video_folder_path).glob("*.mp4"))
        operations += [CommitOperationAdd(path_or_fileobj=str(file), path_in_repo=file) for file in video_files]
        # Push latest one in root directory
        latest_file = max(video_files, key=lambda x: int("".join(filter(str.isdigit, file.stem))))
        operations.append(
            CommitOperationAdd(path_or_fileobj=str(latest_file), path_in_repo=HUGGINGFACE_VIDEO_PREVIEW_FILE_NAME)
        )
```

Only a nit but here you could avoid to sort the full list (only taking `max` is fine, right ?).

vwxyzjn replied:
Nice catch! We only need to the `max`

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,128 @@
+        operations += [CommitOperationAdd(path_or_fileobj=file, path_in_repo=HUGGINGFACE_VIDEO_PREVIEW_FILE_NAME)]
+
+    # fetch folder files
+    for item in [str(item) for item in Path(folder_path).glob("*")]:
+        operations += [CommitOperationAdd(path_or_fileobj=item, path_in_repo=os.path.relpath(item, folder_path))]
```
```suggestion
    operations += [
        CommitOperationAdd(path_or_fileobj=str(item), path_in_repo=item.relative_to(folder_path))
        for item in Path(folder_path).glob("*")
    ]
```

(also a nit): good case here for a list comprehension

(and sorry again for the `str(item)` -> next version of `huggingface_hub` will not require to format the `Path` object)

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,128 @@
+        operations += [CommitOperationAdd(path_or_fileobj=item, path_in_repo=os.path.relpath(item, folder_path))]
+
+    # fetch source code
+    operations += [CommitOperationAdd(path_or_fileobj=sys.argv[0], path_in_repo=sys.argv[0].split("/")[-1])]
```
```suggestion
    operations.append(CommitOperationAdd(path_or_fileobj=sys.argv[0], path_in_repo=sys.argv[0].split("/")[-1]))
```

(nit) concatenating 2 lists makes a copy in memory. Appending an item is not.

vwxyzjn replied:
Oh interesting! Thanks for the info.

Wauplin commented on cleanrl_utils/huggingface.py:
```diff
@@ -0,0 +1,128 @@
+
+    # upload poetry files at the root of the repository
+    git_root = Path(__file__).parent.parent
+    operations += [CommitOperationAdd(path_or_fileobj=f"{git_root}/pyproject.toml", path_in_repo=f"pyproject.toml")]
+    operations += [CommitOperationAdd(path_or_fileobj=f"{git_root}/poetry.lock", path_in_repo=f"poetry.lock")]
```
```suggestion
    operations.append(CommitOperationAdd(path_or_fileobj=str(git_root / "pyproject.toml"), path_in_repo=f"pyproject.toml"))
    operations.append(CommitOperationAdd(path_or_fileobj=str(git_root / "poetry.lock"), path_in_repo=f"poetry.lock"))
```
`f"{git_root}/pyproject.toml"` would not work for Windows users

Wauplin reviewed (commented):
Hey @vwxyzjn, I made another review of the integration. Looks great to me from a `huggingface_hub` standpoint. It's a good way of handling commit operations. I made a few comments, most of them are purely cosmetic. The only real issue I've spotted is with paths on Windows filesystems.

vwxyzjn commented:
Got an example notebook working for all the supported models https://colab.research.google.com/drive/1vhbb4ak9smE7-LiHdaxwTACi1oOC5a4l?usp=sharing#scrollTo=TrQae62Y70H0

<img width="774" alt="image" src="https://user-images.githubusercontent.com/5555347/208781420-a6771136-dff1-4c93-af4b-f92d35deaf78.png">

simoninithomas commented:
Amazing news thanks @vwxyzjn ! 🔥 .

So if we recap:
- For running on your own machine or VM we follow this process: https://github.com/vwxyzjn/cleanrl/blob/hf-integration/docs/get-started/zoo.md.
- On colab we follow the colab process you've just shown.

From my side:
1. To track the number of downloads of a model (displayed in the right widget on model cards) we can keep track of a file. I was thinking of the pickle *pth 
2. I'm adding CleanRL to the official CleanRL in the tags with library_name: cleanrl. Given it's holiday season this might be merged only next week.
3. I'll update Hub docs with the nice doc you've already done and make some cleanups.

🤗

vwxyzjn commented:
Sounds good on my end. We should aim for a release after the new year. Happy holiday all :)

vwxyzjn commented:
Merging this as is, subjecting to future PRs. We'd also probably use https://github.com/huggingface/blog/pull/616 to make the announcement. Thanks for the great work, folks!

Wauplin commented:
Congrats ! That was a big piece of work 🎉🎉

simoninithomas commented:
Congratulations 👏  I was off at the end of last week. I'm prepare the blogpost for next week and we're going to have a unit using CleanRL on PPO with Edward and me using GodotRL we will have the PR this week I'll mention you to put you in the loop.
